diff --git a/drivers/net/rtl8169.c b/drivers/net/rtl8169.c
index 3896760b5f..b534e2a65c 100644
--- a/drivers/net/rtl8169.c
+++ b/drivers/net/rtl8169.c
@@ -99,12 +99,12 @@ static int media[MAX_UNITS] = { -1, -1, -1, -1, -1, -1, -1, -1 };
 #define TX_TIMEOUT  (6*HZ)
 
 /* write/read MMIO register. Notice: {read,write}[wl] do the necessary swapping */
-#define RTL_W8(reg, val8)	writeb((val8), ioaddr + (reg))
-#define RTL_W16(reg, val16)	writew((val16), ioaddr + (reg))
-#define RTL_W32(reg, val32)	writel((val32), ioaddr + (reg))
-#define RTL_R8(reg)		readb(ioaddr + (reg))
-#define RTL_R16(reg)		readw(ioaddr + (reg))
-#define RTL_R32(reg)		readl(ioaddr + (reg))
+#define RTL_W8(reg, val8)	writeb((val8), (void*) ioaddr + (reg))
+#define RTL_W16(reg, val16)	writew((val16), (void*) ioaddr + (reg))
+#define RTL_W32(reg, val32)	writel((val32), (void*) ioaddr + (reg))
+#define RTL_R8(reg)		readb((void*) ioaddr + (reg))
+#define RTL_R16(reg)		readw((void*) ioaddr + (reg))
+#define RTL_R32(reg)		readl((void*) ioaddr + (reg))
 
 #define bus_to_phys(a)	pci_mem_to_phys((pci_dev_t)(unsigned long)dev->priv, \
 	(pci_addr_t)(unsigned long)a)
diff --git a/drivers/usb/cdns3/cdns3-starfive.c b/drivers/usb/cdns3/cdns3-starfive.c
index 631fae21cd..9a48ee6a04 100644
--- a/drivers/usb/cdns3/cdns3-starfive.c
+++ b/drivers/usb/cdns3/cdns3-starfive.c
@@ -11,6 +11,7 @@
 #include <clk.h>
 #include <dm.h>
 #include <dm/device_compat.h>
+#include <dm/lists.h>
 #include <linux/usb/otg.h>
 #include <reset.h>
 #include <regmap.h>
diff --git a/drivers/video/dw_hdmi.c b/drivers/video/dw_hdmi.c
index c4fbb18294..9b498e17e9 100644
--- a/drivers/video/dw_hdmi.c
+++ b/drivers/video/dw_hdmi.c
@@ -13,6 +13,7 @@
 #include <media_bus_format.h>
 #include <linux/delay.h>
 #include "dw_hdmi.h"
+#include <stddef.h>
 
 struct tmds_n_cts {
 	u32 tmds;
@@ -78,10 +79,10 @@ static void dw_hdmi_write(struct dw_hdmi *hdmi, u8 val, int offset)
 {
 	switch (hdmi->reg_io_width) {
 	case 1:
-		writeb(val, hdmi->ioaddr + offset);
+		writeb(val, (void*) hdmi->ioaddr + offset);
 		break;
 	case 4:
-		writel(val, hdmi->ioaddr + (offset << 2));
+		writel(val, (void*) hdmi->ioaddr + (offset << 2));
 		break;
 	default:
 		debug("reg_io_width has unsupported width!\n");
@@ -93,9 +94,9 @@ static u8 dw_hdmi_read(struct dw_hdmi *hdmi, int offset)
 {
 	switch (hdmi->reg_io_width) {
 	case 1:
-		return readb(hdmi->ioaddr + offset);
+		return readb((void*) hdmi->ioaddr + offset);
 	case 4:
-		return readl(hdmi->ioaddr + (offset << 2));
+		return readl((void*) hdmi->ioaddr + (offset << 2));
 	default:
 		debug("reg_io_width has unsupported width!\n");
 		break;
diff --git a/include/elf.h b/include/elf.h
index b04e746d61..7dacae2c87 100644
--- a/include/elf.h
+++ b/include/elf.h
@@ -360,6 +360,15 @@ typedef struct {
 	Elf32_Half	st_shndx;	/* section header index */
 } Elf32_Sym;
 
+typedef struct {
+	Elf64_Word	st_name;	/* name - index into string table */
+	Elf64_Addr	st_value;	/* symbol value */
+	Elf64_Word	st_size;	/* symbol size */
+	unsigned char	st_info;	/* type and binding */
+	unsigned char	st_other;	/* 0 - no defined meaning */
+	Elf64_Half	st_shndx;	/* section header index */
+} Elf64_Sym;
+
 /* Symbol table index */
 #define STN_UNDEF	0		/* undefined */
 
diff --git a/tools/imagetool.c b/tools/imagetool.c
index ba1f64aa37..7ca55c7a33 100644
--- a/tools/imagetool.c
+++ b/tools/imagetool.c
@@ -9,6 +9,10 @@
 
 #include <image.h>
 
+#if defined(__MACH__)
+struct image_type_params **__start_image_type, **__stop_image_type;
+#endif
+
 struct image_type_params *imagetool_get_type(int type)
 {
 	struct image_type_params **curr;
diff --git a/tools/imagetool.h b/tools/imagetool.h
index e229a34ffc..07215fc1d3 100644
--- a/tools/imagetool.h
+++ b/tools/imagetool.h
@@ -283,7 +283,7 @@ int rockchip_copy_image(int fd, struct image_tool_params *mparams);
 	} while (0)
 #define SECTION(name)   __attribute__((section("__DATA, " #name)))
 
-struct image_type_params **__start_image_type, **__stop_image_type;
+extern struct image_type_params **__start_image_type, **__stop_image_type;
 #else
 #define INIT_SECTION(name) /* no-op for ELF */
 #define SECTION(name)   __attribute__((section(#name)))
